### PR TITLE
feat(muon): add Muon optimizer support to the DTensor backend

### LINKED ISSUE #2510: [Feature] Muon optimizer support for the DTensor backend
## Summary

Muon is currently a Megatron-only optimizer in NeMo-RL (`docs/guides/muon-optimizer.md:11`). I'd like to make it available on the DTensor backend too.

## Why

Optimizer choice and backend choice are separable concerns everywhere else. AdamW, schedulers, and gradient clipping all work on both backends. Muon is the only optimizer that gates on backend.

It matters because DTensor is the on-ramp the project itself recommends. All three exemplar configs (`sft.yaml`, `grpo_math_1B.yaml`, `dpo.yaml`) default to `dtensor_cfg.enabled: true`, the README quickstart runs them as-is, and Megatron is the path the README explicitly flags as needing extra setup care (cuDNN env vars, submodule init, TE built from source). For a user trying NeMo-RL for the first time on a small model, the recommended path silently removes Muon as an option even though Muon is exactly the kind of thing a small-scale experimenter would want to try. Its documented wins in `docs/guides/muon-optimizer.md` (Qwen3-235B SFT, Qwen2.5-7B DAPO) are about sample efficiency, which is precisely what matters at small scale.

### PR DESCRIPTION
**Pull request by bzantium**
# What does this PR do ?

Brings the Muon optimizer (MomentUm Orthogonalized by Newton-Schulz, https://arxiv.org/abs/2502.16982) to the DTensor backend. `docs/guides/muon-optimizer.md:11` previously documented the optimizer as Megatron-only; this PR lifts that restriction so users on FSDP2 / TP recipes can train with Muon without switching backends.

The optimizer math is reused verbatim from NVIDIA's `emerging_optimizers` package (the same one Megatron's `dist_muon` shims over). The DTensor side is a thin adapter: a placement to `partition_dim` translator, an all-gather path for `tp_mode="duplicated"`, and `newton_schulz_tp` for the distributed path. Equivalence is verified to bit-exact (`atol=0, rtol=0`) vs upstream `emerging_optimizers.Muon` on single-GPU, bit-exact for TP=2 `duplicated`/`blockwise`, within 5e-4 for TP=2 `distributed`, and max |Δ| 1.4e-3 across a 100-step Llama-3.2-1B SFT smoke vs Megatron `dist_muon` on identical recipe. The Megatron `dist_muon` path and existing AdamW recipes are untouched; the dispatch is name-based in `nemo_rl/utils/optimizer_factory.py`.

# Issues

Closes #2510.

# Usage

```yaml
policy:
  optimizer:
    name: nemo_rl.algorithms.muon.build_dtensor_muon
    kwargs:
      lr: 2.0e-5
      weight_decay: 0.1
      muon_momentum: 0.9
      muon_nesterov: false
      muon_scale_mode: spectral
      muon_extra_scale_factor: 0.2
      muon_tp_mode: duplicated   # blockwise | duplicated | distributed
      muon_split_qkv: true
      adamw_kwargs:
        betas: [0.9, 0.999]
        eps: 1.0e-8
```

`uv sync --extra automodel` brings in `emerging-optimizers` for DTensor users; the existing `mcore` extras still ship the same package for the Megatron path.

# Before your PR is \"Ready for review\"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [x] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information

* Optimizer config schema is unchanged (`PytorchOptimizerConfig` already accepts an open `kwargs` dict); no hidden defaults in code, and the exemplar recipe carries the explicit knob set.
* The factory uses `importlib` directly rather than `hydra.utils.get_class` because `get_class` rejects callables that are not classes, which would prevent dispatching to the `build_dtensor_muon` factory function.

### CODE REVIEW

**Comment by bzantium** [comment]
## Megatron `dist_muon` vs DTensor `build_dtensor_muon` equivalence

Single-node 8xA100 SFT run, identical recipe except for the optimizer construction path (same model, dataset, batch shape, seed, hyperparameters).

| | Megatron `dist_muon` | DTensor `build_dtensor_muon` |
| --- | --- | --- |
| Model | meta-llama/Llama-3.2-1B-Instruct | meta-llama/Llama-3.2-1B-Instruct |
| Dataset | OpenMathInstruct-2 (split_validation_size 0.05) | (same) |
| Steps / GBS / MBS / seq | 100 / 32 / 2 / 2048 | (same) |
| Optimizer knobs | lr 2e-5, wd 0.1, momentum 0.9, scale=spectral, extra_scale_factor=0.2, split_qkv | (identical), `tp_mode=duplicated` |
| Train loss @ step 100 | **0.4355** | **0.4350** |
| Validation loss @ step 50 | **0.4961** | **0.4957** |
| max \|Δ\| over 100 steps | | **0.0014** |
| mean \|Δ\| over 100 steps | | **0.0004** |

Δ stays within FP non-determinism for bf16 matmul and Newton-Schulz orthogonalization across the two backends, which confirms the DTensor builder reproduces the Megatron `dist_muon` math.

Muon equivalence loss curves: Megatron dist_muon vs DTensor build_dtensor_muon

Top panel: train loss for both backends (mcore solid blue, dtensor dashed red). Bottom panel: per-step delta (`dtensor − mcore`).

Throughput is not a controlled comparison here (mcore ~144 TFLOPS, dtensor ~75 TFLOPS at this shape). The DTensor backend's per-step cost is dominated by HuggingFace model-fwd rather than Muon orthogonalization, so closing the gap is orthogonal to the equivalence check this PR is about and belongs in a separate follow-up.
